cpp/include/raft/comms/std_comms.hpp

```diff
@@ -0,0 +1,406 @@
+
+    CUDA_CHECK(cudaMalloc(&_sendbuff, sizeof(int)));
+    CUDA_CHECK(cudaMalloc(&_recvbuff, sizeof(int)));
+  }
+
+  int getSize() const { return _size; }
+
+  int getRank() const { return _rank; }
+
+  std::unique_ptr<comms_iface> commSplit(int color, int key) const {
+    // Not supported by NCCL
+    ASSERT(false,
+           "ERROR: commSplit called but not yet supported in this comms "
+           "implementation.");
+  }
+
+  void barrier() const {
+    CUDA_CHECK(cudaMemsetAsync(_sendbuff, 1, sizeof(int), _stream));
+    CUDA_CHECK(cudaMemsetAsync(_recvbuff, 1, sizeof(int), _stream));
+
+    allreduce(_sendbuff, _recvbuff, 1, INT, SUM, _stream);
+
+    ASSERT(syncStream(_stream) == status_t::commStatusSuccess,
+           "ERROR: syncStream failed. This can be caused by a failed rank.");
+  }
+
+  void get_request_id(request_t *req) const {
+    request_t req_id;
+
+    if (this->_free_requests.empty())
+      req_id = this->_next_request_id++;
+    else {
+      auto it = this->_free_requests.begin();
+      req_id = *it;
+      this->_free_requests.erase(it);
+    }
+    *req = req_id;
+  }
+
+  void isend(const void *buf, int size, int dest, int tag,
+             request_t *request) const {
+    ASSERT(_ucp_worker != nullptr,
+           "ERROR: UCX comms not initialized on communicator.");
+
+    get_request_id(request);
+    ucp_ep_h ep_ptr = (*_ucp_eps)[dest];
+
+    ucp_request *ucp_req = (ucp_request *)malloc(sizeof(ucp_request));
+
+    this->_ucp_handler.ucp_isend(ucp_req, ep_ptr, buf, size, tag,
+                                 default_tag_mask, getRank());
+
+    _requests_in_flight.insert(std::make_pair(*request, ucp_req));
+  }
+
+  void irecv(void *buf, int size, int source, int tag,
+             request_t *request) const {
+    ASSERT(_ucp_worker != nullptr,
+           "ERROR: UCX comms not initialized on communicator.");
+
+    get_request_id(request);
+
+    ucp_ep_h ep_ptr = (*_ucp_eps)[source];
+
+    ucp_tag_t tag_mask = default_tag_mask;
+
+    ucp_request *ucp_req = (ucp_request *)malloc(sizeof(ucp_request));
+    _ucp_handler.ucp_irecv(ucp_req, _ucp_worker, ep_ptr, buf, size, tag,
+                           tag_mask, source);
+
+    _requests_in_flight.insert(std::make_pair(*request, ucp_req));
+  }
+
+  void waitall(int count, request_t array_of_requests[]) const {
+    ASSERT(_ucp_worker != nullptr,
+           "ERROR: UCX comms not initialized on communicator.");
+
+    std::vector<ucp_request *> requests;
+    requests.reserve(count);
+
+    time_t start = time(NULL);
+
+    for (int i = 0; i < count; ++i) {
+      auto req_it = _requests_in_flight.find(array_of_requests[i]);
+      ASSERT(_requests_in_flight.end() != req_it,
+             "ERROR: waitall on invalid request: %d", array_of_requests[i]);
+      requests.push_back(req_it->second);
+      _free_requests.insert(req_it->first);
+      _requests_in_flight.erase(req_it);
+    }
+
+    while (requests.size() > 0) {
```

Comment:
and UCX has no MPI_Waitall kinda function? Should we implement this by ourselves?

Reply:
UCX is a layer below MPI. This is at the UCP layer, which only knows endpoints and asynchronous message receiving/callbacks. It’s up to the user to implement blocking functions for multiple messages.

Reply:
I see, there is MPI libraries with UCX backend, and they may have exhaustively tuned their code for this, but IIRC, we have a reason we cannot use MPI, right? Little uneasy about implementing this by ourselves, but I guess we cannot avoid this to move forward :-)